src/Data/Bin/Base.agda

```diff
@@ -0,0 +1,129 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Arithmetic of binary represented natural numbers. Initial part.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Bin.Base where
+
+import      Algebra.FunctionProperties as FuncProp
+open import Data.Nat using (ℕ) renaming (suc to 1+_; _*_ to _*ₙ_)
+open import Function        using (_∘_)
+open import Relation.Binary using (Decidable)
+open import Relation.Binary.PropositionalEquality using (_≡_; refl; cong)
+open import Relation.Nullary                      using (yes; no)
+
+
+data Bin : Set
+  where
+  0#    : Bin
+  2suc  : Bin → Bin    -- \n→ 2*(1+n)  arbitrary nonzero even
+  suc2* : Bin → Bin    -- \n→ 1 + 2n   arbitrary odd
+--
+-- This representation is unique for each natural number.
+
+
+open FuncProp (_≡_ {A = Bin}) using (Op₂)
+
+2suc-injective : {x y : Bin} → 2suc x ≡ 2suc y → x ≡ y
+2suc-injective {0#}      {0#}      _    =  refl
+2suc-injective {2suc _}  {2suc _}  refl =  refl
+2suc-injective {suc2* _} {suc2* _} refl =  refl
+
+suc2*-injective : {x y : Bin} → suc2* x ≡ suc2* y → x ≡ y
+suc2*-injective {0#}      {0#}      _    =  refl
+suc2*-injective {2suc _}  {2suc _}  refl =  refl
+suc2*-injective {suc2* _} {suc2* _} refl =  refl
+
+-- Decidable equality on Bin.
+
+_≟_ :  Decidable (_≡_ {A = Bin})
+
```

Comment:
Please don't leave blank spaces between proofs, unless very large. Otherwise they end up looking like multiple proofs.

Reply:
Done (may be not everywhere).